lib/LaTeXML/Core/Gullet.pm

```diff
@@ -535,30 +531,51 @@ sub readKeyword {
   return; }
 
 # Return a (balanced) sequence tokens until a match against one of the Tokens in @delims.
-# In list context, also returns the found delimiter.
+# Note that Braces on input hides the contents from matching,
+# so this assumes there wont be braces in $delim!
+# But, see readUntilBrace for that case.
+# NOTE: Add error handling, $token undef detection, etc!!!!!
 sub readUntil {
-  my ($self, @delims) = @_;
-  my ($n, $found, @tokens) = (0);
-  while (!defined($found = $self->readMatch(@delims))) {
-    my $token = $self->readToken();    # Copy next token to args
-###    return unless defined $token;
-    if (!defined $token) {             # Ran out!
-      print STDERR "UNTIL failed!\n";
-      print STDERR "Seeking one of " .
-        join(' or ', map { join(' ', map { Stringify($_) } $_->unlist); } @delims) . "\n";
-      print STDERR "Read so far: " . join(' ', map { Stringify($_); } @tokens) . "\n";
-      # Not more correct, but maybe less confusing if we put the read tokens BACK????
-      $self->unread(@tokens);
-      return; }
-    push(@tokens, $token);
-    $n++;
-    if ($$token[1] == CC_BEGIN) {      # And if it's a BEGIN, copy till balanced END
-      push(@tokens, $self->readBalanced); } }
+  my ($self, $delim) = @_;
+  my @tokens = ();
+  my $token;
+  my $nbraces  = 0;
+  my @want     = $delim->unlist;
+  my $ntomatch = scalar(@want);
+  if ($ntomatch == 1) {    # Common, easy case: read till we match a single token
+    my $want = $want[0];
+    #    while(($token = $self->readToken) && !$token->equals($want)){
+    while (($token = shift(@{ $$self{pushback} }) || $$self{mouth}->readToken())
+      && (($$token[1] != CC_SMUGGLE_THE) || ($token = $$token[2]))
+      && !$token->equals($want)) {
+      push(@tokens, $token);
+      if ($$token[1] == CC_BEGIN) {    # And if it's a BEGIN, copy till balanced END
+        $nbraces++;
+        push(@tokens, $self->readBalanced, T_END); } } }
+  else {
+    my @ring = ();
+    while (1) {
+      # prefill the required number of tokens
+      while (scalar(@ring) < $ntomatch) {
+        $token = $self->readToken;
+        if ($$token[1] == CC_BEGIN) {    # read balanced, and refill ring.
+          $nbraces++;
+          push(@tokens, @ring, $token, $self->readBalanced, T_END);    # Copy directly to result
+          @ring = (); }                                                # and retry
+        else {
+          push(@ring, $token); } }
+      my $i;
+      for ($i = 0 ; ($i < $ntomatch) && ($ring[$i]->equals($want[$i])) ; $i++) { }    # Test match
+      last if $i >= $ntomatch;                                                        # Matched all!
+      push(@tokens, shift(@ring)); } }
+  if (!defined $token) {                                                              # Ran out!
+    $self->unread(@tokens);    # Not more correct, but maybe less confusing?
+    return; }
```

Comment:
A bit scary that `readUntil` got a refactor of this size. Is the new code closer to TeX, or ideally functional identical to the old?

Reply:
Yeah, scary perhaps, 'cause it is a change to api to conform with the way it's actually used (which then allowed for some nice optimization).